backend/pikpak/pikpak.go

```diff
@@ -0,0 +1,1707 @@
+	// the above code is no longer sufficient to handle the 'invalid_grant' error.
+	if strings.Contains(err.Error(), "invalid_grant") {
+		authRetry = true
+	}
+
+	if authRetry {
+		if authErr := f.reAuthorize(ctx); authErr != nil {
+			return false, fserrors.FatalError(authErr)
+		}
+	}
+
+	switch apiErr := err.(type) {
+	case *api.Error:
+		if apiErr.Reason == "file_rename_uncompleted" {
+			// "file_rename_uncompleted" (9): Renaming uncompleted file or folder is not supported
+			// This error occurs when you attempt to rename objects
+			// right after some server-side changes, e.g. DirMove, Move, Copy
+			return true, err
+		} else if apiErr.Reason == "file_duplicated_name" {
+			// "file_duplicated_name" (3): File name cannot be repeated
+			// This error may occur when attempting to rename temp object (newly uploaded)
+			// right after the old one is removed.
+			return true, err
+		} else if apiErr.Reason == "task_daily_create_limit_vip" {
+			// "task_daily_create_limit_vip" (11): Sorry, you have submitted too many tasks and have exceeded the current processing capacity, please try again tomorrow
+			return false, fserrors.FatalError(err)
+		} else if apiErr.Reason == "file_space_not_enough" {
+			// "file_space_not_enough" (8): Storage space is not enough
+			return false, fserrors.FatalError(err)
+		}
+	}
+
+	return authRetry || fserrors.ShouldRetryHTTP(resp, retryErrorCodes), err
+}
+
+// errorHandler parses a non 2xx error response into an error
+func errorHandler(resp *http.Response) error {
+	// Decode error response
+	errResponse := new(api.Error)
+	err := rest.DecodeJSON(resp, &errResponse)
+	if err != nil {
+		fs.Debugf(nil, "Couldn't decode error response: %v", err)
+	}
+	if errResponse.Reason == "" {
+		errResponse.Reason = resp.Status
+	}
+	if errResponse.Code == 0 {
+		errResponse.Code = resp.StatusCode
+	}
+	return errResponse
+}
+
+// newClientWithPacer sets a new http/rest client with a pacer to Fs
+func (f *Fs) newClientWithPacer(ctx context.Context) (err error) {
+	f.client, _, err = oauthutil.NewClient(ctx, f.name, f.m, oauthConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create oauth client: %w", err)
+	}
+	f.rst = rest.NewClient(f.client).SetRoot(rootURL).SetErrorHandler(errorHandler)
+	f.pacer = fs.NewPacer(ctx, pacer.NewDefault(pacer.MinSleep(minSleep), pacer.MaxSleep(maxSleep), pacer.DecayConstant(decayConstant)))
+	return nil
+}
+
+// newFs partially constructs Fs from the path
+//
+// It constructs a valid Fs but doesn't attempt to figure out whether
+// it is a file or a directory.
+func newFs(ctx context.Context, name, path string, m configmap.Mapper) (*Fs, error) {
+	// Parse config into Options struct
+	opt := new(Options)
+	if err := configstruct.Set(m, opt); err != nil {
+		return nil, err
+	}
+
+	root := parsePath(path)
+
+	// overrides global `--multi-thread-streams` by local one
+	ci := fs.GetConfig(ctx)
+	ci.MultiThreadStreams = opt.MultiThreadStreams
```

Comment:
I don't like this very much! I'd rather it was a Feature we advertise and get the multithread streaming code to adapt.

However its OK for the initial merge - we can fixup later.